### ISSUE TITLE
Regression on regex when upgrading from 0.6.28 to 0.6.31

### ISSUE THREAD

**Issue by dportabella** [scala-js/scala-js#3901]
```
// project/plugins.sbt: addSbtPlugin("org.scala-js" % "sbt-scalajs" % "0.6.31")
import scala.util.matching.Regex
val text = ???
val regex: String = ???
val it: Iterator[Regex.Match] = regex.r.findAllMatchIn(text)
println(it.hasNext)  // returns true
it.next // it fails with:

TypeError: Cannot read property 'length' of null
    at $c_ju_regex_Matcher.end__I__I (RuntimeString.scala:252)
    at $c_s_util_matching_Regex$Match.ends$lzycompute__p1__AI (Regex.scala:732)
    at $c_s_util_matching_Regex$Match.ends__p1__AI (Regex.scala:731)
    at $c_s_util_matching_Regex$Match.force__s_util_matching_Regex$Match (Regex.scala:743)
    at $c_s_util_matching_Regex$$anon$1.next__s_util_matching_Regex$Match (Regex.scala:410)
    at $c_s_util_matching_Regex$$anon$1.next__O (Regex.scala:406)
```

it works by changing sbt-scalajs version from 0.6.31 to 0.6.28.

Unfortunately I cannot give the text and regex as it is proprietary and very long.
I'll try to minimize the text and regex and share it here when I have some time.

**Comment by sjrd** [comment]
Thank you for the report.

What version of Scala are you using?

**Comment by dportabella** [comment]
scala 2.12.10

**Comment by sjrd** [comment]
This was most certainly caused by https://github.com/scala-js/scala-js/commit/56dd7abb9c32ae519c8327ee574295687ca96127.

The error is basically a `NullPointerException` that is a consequence of `start(group)` not returning -1 where before it correctly returned -1 at
https://github.com/scala-js/scala-js/blob/3d0e4e3f39644218c5fa6bf180d653b9bb33495a/javalib/src/main/scala/java/util/regex/Matcher.scala#L199
Computing `start(group)` is the whole point of `GroupStartMapper`, so something went wrong there.

@dportabella A reproduction would be really welcome. One hint I can give is that, if my above diagnosis is correct, this has to do with a match that does match something as a whole, but one group inside does not match anything (it is in a non-chosen alternative with `|`, or in a non-present optional thing with `?`).

**Comment by sjrd** [comment]
Here is a minimized reproduction:
```scala
    val text = "b"
    val regex = "(?!(a))b"
    val pattern = java.util.regex.Pattern.compile(regex)
    val matcher = pattern.matcher(text)
    println(matcher.find())
    println(matcher.start(1)) // returns 0 instead of -1
    println(matcher.end(1)) // throws
    val it = regex.r.findAllMatchIn(text)
    while (it.hasNext)
      println(it.next()) // throws
```
It happens with a capturing group within a negative look-ahead.

Fix incoming.

**Comment by sjrd** [comment]
Fixed in 602005dedc8375be33887774271bace1c851615c.
